firebase_admin/_token_gen.py

```diff
@@ -254,98 +275,134 @@ def __init__(self, **kwargs):
+
+    def _verify_sub(self, payload):
+        subject = payload.get('sub')
+        if subject is None or not isinstance(subject, six.string_types):
+            raise self._invalid_token_error(
                 'Firebase {0} has no "sub" (subject) claim. '
-                '{1}'.format(self.short_name, verify_id_token_msg))
+                '{1}'.format(self.short_name, self.verify_id_token_msg))
         elif not subject:
-            error_message = (
+            raise self._invalid_token_error(
                 'Firebase {0} has an empty string "sub" (subject) claim. '
-                '{1}'.format(self.short_name, verify_id_token_msg))
+                '{1}'.format(self.short_name, self.verify_id_token_msg))
         elif len(subject) > 128:
-            error_message = (
+            raise self._invalid_token_error(
                 'Firebase {0} has a "sub" (subject) claim longer than 128 characters. '
-                '{1}'.format(self.short_name, verify_id_token_msg))
-
-        if error_message:
-            raise self._invalid_token_error(error_message)
-
+                '{1}'.format(self.short_name, self.verify_id_token_msg))
+
+    def _verify_iat_and_exp(self, payload):
+        issued_at = payload.get('iat')
+        expires_at = payload.get('exp')
+        now = calendar.timegm(datetime.datetime.utcnow().utctimetuple())
+        if not issued_at:
+            raise self._invalid_token_error('Token does not contain required claim "iat"')
+        elif not expires_at:
+            raise self._invalid_token_error('Token does not contain required claim "exp"')
+        elif now < issued_at - _CLOCK_SKEW_SECS:
+            raise self._invalid_token_error(
+                'Token used too early, {0} < {1}'.format(now, issued_at))
+        elif now > expires_at + _CLOCK_SKEW_SECS:
+            raise self._expired_token_error(
+                'Token expired, {0} < {1}'.format(expires_at, now), cause=None)
+
+    def _verify_token_signature(self, token, request):
         try:
-            verified_claims = google.oauth2.id_token.verify_token(
-                token,
-                request=request,
-                audience=self.project_id,
-                certs_url=self.cert_url)
-            verified_claims['uid'] = verified_claims['sub']
-            return verified_claims
+            return google.oauth2.id_token.verify_token(
+                token, request=request, audience=self.project_id, certs_url=self.cert_url)
         except google.auth.exceptions.TransportError as error:
             raise CertificateFetchError(str(error), cause=error)
         except ValueError as error:
-            if 'Token expired' in str(error):
-                raise self._expired_token_error(str(error), cause=error)
-            raise self._invalid_token_error(str(error), cause=error)
+            error_factory = (self._expired_token_error if 'Token expired' in str(error) else
```

Comment:
Nit: Can afford to leave this out of the diff.

Reply:
Done